src/tracing/agent.cc

```diff
@@ -1,67 +1,105 @@
 #include "tracing/agent.h"
-
-#include <sstream>
-#include <string>
 #include "tracing/node_trace_buffer.h"
 #include "tracing/node_trace_writer.h"
-
 #include "env-inl.h"
 
+#include <sstream>
+#include <string>
+#include <vector>
+
 namespace node {
 namespace tracing {
 
 using v8::platform::tracing::TraceConfig;
-using std::string;
 
-Agent::Agent(const std::string& log_file_pattern) {
+Agent::Agent(const std::string& log_file_pattern) :
+    log_file_pattern_(log_file_pattern) {
+  tracing_controller_ = new TracingController();
+  tracing_controller_->Initialize(nullptr);
+}
+
+void Agent::InitializeOnce() {
+  if (initialized_)
+    return;
   int err = uv_loop_init(&tracing_loop_);
   CHECK_EQ(err, 0);
 
   NodeTraceWriter* trace_writer = new NodeTraceWriter(
-      log_file_pattern, &tracing_loop_);
+      log_file_pattern_, &tracing_loop_);
   TraceBuffer* trace_buffer = new NodeTraceBuffer(
       NodeTraceBuffer::kBufferChunks, trace_writer, &tracing_loop_);
-  tracing_controller_ = new TracingController();
   tracing_controller_->Initialize(trace_buffer);
 }
 
-void Agent::Start(const string& enabled_categories) {
-  TraceConfig* trace_config = new TraceConfig();
-  if (!enabled_categories.empty()) {
-    std::stringstream category_list(enabled_categories);
+void Agent::StartTracing() {
+  tracing_controller_->StopTracing();
+  if (!categories_.empty()) {
+    if (!started_) {
+      InitializeOnce();
+      // This thread should be created *after* async handles are created
+      // (within NodeTraceWriter and NodeTraceBuffer constructors).
+      // Otherwise the thread could shut down prematurely.
+      int err = uv_thread_create(&thread_, ThreadCb, this);
+      CHECK_EQ(err, 0);
+      started_ = true;
+    }
+    TraceConfig* trace_config = new TraceConfig();
+    for (auto it = categories_.begin(); it != categories_.end(); it++) {
```

Comment:
Ditto – `for (const std::string& category : categories)` would read better, imo.